Fixed CRGB bug

MyloWhylo opened the pull request:
This fixes a few problems:

First, it resolves a bug related to trying to pass a `CRGB` object into a function that expects a color in `RGB565` format. I wrote a simple bit manipulation that puts all the relevant bits into their proper `uint16_t` places, resolving this bug.

Second, it resolves a few compiler warnings about unused variables, by marking them as unused.

Third, it adds proper Platform.IO information to the `.gitignore`.

Fourth, uhhh idk I didn't really mean to change the `platformio.ini` configuration, but it now uses a Windows serial port object, which is interesting.

MyloWhylo committed:
Fixed CRGB bug

MyloWhylo commented:
Fixes #1

b0xcat commented on platformio.ini:
```diff
@@ -19,8 +19,8 @@ board_build.filesystem = littlefs
 extra_scripts = replace_fs.py
 upload_speed = 921600
 monitor_speed = 115200
-upload_port = /dev/ttyUSB0
```
I think maybe we can get rid of `monitor_port` and `upload_port` here altogether, since they're gonna be different for different people anyways

MyloWhylo replied:
Yep, I agree with that! If it's possible to leave that out of the ini file, I think that'd be the way to go.

b0xcat commented on src/scene/visitors/elementrgbbitmapsetter.cpp:
```diff
@@ -9,6 +9,11 @@
 #include "bitmaps.h"
 
 
+inline uint16_t crgbToRGB565(CRGB color) {
```
Color conversion functions are already defined in `include/colorconversion.h`, no need to define one here. This exact one is on line 46.

b0xcat commented on src/scene/visitors/elementrgbbitmapsetter.cpp:
```diff
@@ -26,7 +31,7 @@ void ElementRGBBitmapSetter::visit(AdafruitGFXElement* el) {
         for (uint16_t x = 0; x < el->width(); x++) {
             for (uint16_t y = 0; y < el->height(); y++) {
-                el->drawPixel(x, y, cur_bitmap->getPixel(x, y));
+                el->drawPixel(x, y, crgbToRGB565(cur_bitmap->getPixel(x, y)));
```
Good catch here! I hadn't used the adafruit elements myself in a while so I didn't know they were broken! (I need to set up some tests at some point I guess)

b0xcat reviewed (commented):
Hey Mylo, thanks for taking a look at this! I reviewed your changes with some comments. If you feel like doing that stuff, feel free, if not let me know and I'll do it!

MyloWhylo committed:
Use alrready-supplied function

MyloWhylo committed:
Made requested changes & fixed bugs

MyloWhylo commented:
OK, I made the requested changes. I don't think that the convert function needs to be `CRGB&` , as we're not modifying the CRGB object, so I changed that. I also noticed some other unused variables, and fixed the `swapped` variable not actually being swapped. Finally, I also resolved some warnings about there being "no hardware spi pins" on esp32 by defining something. It still sends warnings, but they're not actually warnings, which sucks. Please let me know what you think!

b0xcat commented:
Nice, thank you! I was passing in the CRGB struct by reference to prevent it from being copied, but since it is only 3 bytes in size it really doesn't matter, and generates the same assembly either way: https://godbolt.org/z/W31v4xY7r

Good catch on the `swapped` variable, it was something I was working on a while ago (while too eepy apparently lol) but never got around to finishing. Thanks!

I looked into the fastled warnings, and apparently there's not too much we can do about that so we'll just have to live with them. It doesn't really bother me so no big deal.

Oh and before merging I guess I should check if you're cool with me changing the license from GPLv3 to AGPLv3

If that's all good I'll merge!

MyloWhylo commented:
I don't have any issues with that, I honestly don't know enough about AGPL to really be able to comment on it lol. And the reason I was passing by value instead of by reference is that it wasn't wanting to compile, and I didn't want to deal with pointer stuff lol. If you think it's good to merge, I'd say send it.
